62.SchusslerTest/shader.frag

```diff
@@ -0,0 +1,127 @@
+#version 430 core
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+
+layout (location = 0) in vec3 Normal;
+layout (location = 1) in vec3 Pos;
+layout (location = 2) flat in vec3 LightPos;
+layout (location = 3) flat in vec3 CamPos;
+layout (location = 4) flat in vec3 GNormal;
+layout (location = 5) flat in float Alpha;
+layout (location = 6) flat in float AlphaY;
+
+layout (location = 0) out vec4 outColor;
+
+layout (push_constant) uniform PC {
+    layout (offset = 64) vec3 campos;
+    layout (offset = 80) uint testNum;
+} pc;
+
+#define TEST_GGX 1
+#define TEST_BECKMANN 2
+#define TEST_PHONG 3
+#define TEST_AS 4
+#define TEST_OREN_NAYAR 5
+#define TEST_LAMBERT 6
+
+#include <nbl/builtin/glsl/bxdf/brdf/specular/ggx.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/specular/beckmann.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/specular/blinn_phong.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/diffuse/oren_nayar.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/diffuse/lambert.glsl>
+
+float pdot(vec3 a, vec3 b) {
+    return max(0.0, dot(a,b));
+}
+
+float ap(vec3 wi, vec3 wp, vec3 wg) {
+    return pdot(wi, wp)/pdot(wp, wg);
+}
+
+float at(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return pdot(wi, wt) * sqrt(1-(pdot(wp,wg)*pdot(wp,wg))) / pdot(wp,wg);
+}
+
+float lambdap(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return ap(wi, wp, wg) / (ap(wi,wp,wg) + at(wi,wt,wp,wg));
+}
+
+float lambdat(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return at(wi,wt,wp,wg) / (ap(wi,wp,wg) + at(wi,wt,wp,wg));
+}
+
+float G1(vec3 wo, vec3 wp, float a2) {
+    float dotNV = pdot(wp,wo);
+    float denomC = sqrt(a2 + (1.0f - a2) * dotNV * dotNV) + dotNV;
+
+    return 2.0f * dotNV / denomC;
+}
```

Comment:
your G1 is completely wrong, it should be
```glsl
// `max` means its already after `max(,0.f)` treatment
// `X` means either L or V depending on whether you evaluate with `w_o` or `w_i
float schussler_lambda_den(float maxShadingNdotX, float maxSpecularTdotX)
{
   return maxShadingNdotX+maxSpecularTdotX;
}

// [0] lambda_p
// [1] lambda_t
vec2 lambda(float maxShadingNdotX, float lambda_den)
{
   const float lambda_p = maxShadingNdotX/lambda_den;
   return vec2(lambda_p,1.f-lambda_p);
}

// in Nabla, the half vector is denoted `H`, which is `w_m` in the Schussler paper
float G1(float HdotX, float GeomNdotX, float GeomNdotShadingN/*old deviation*/, float lambda_den)
{
   // microfacet cannot be facing away from the direction, which is why the Heaviside function was in Eq 13
   if (HdotX>0.f)
   {
      const float rcp_a_sum = GeomNdotShadingN/lambda_den;
      return min(GeomNdotX*rcp_a_sum,1.f);
   }
   return 0.f;
}
```

Reply:
Your G2 becomes
```glsl
// even though we call `w_m` `H` its not a half vector, because no perfect specularity happens
float schussler_G2(float HdotV, float GeomNdotV, float NdotV, float HdotL, float GeomNdotL, float NdotL, float lambda_den)
{
   return schussler_G1(HdotV,GeomNdotV,NdotV,HdotL,GeomNDotL,NdotL,lambda_den);
}
```

Reply:
the G2 I quoted is wrong, should be
```glsl
float schussler_G2(float HdotV, float GdotV, float rcp_a_sumV, float HdotL, float GdotL, float rcp_a_sumL)
{
   return schussler_G1(HdotV,GdotV,rcp_a_sumV)*schussler_G1(HdotL,GdotL,rcp_a_sumL);
}
```
the `rcp_a_sum` is different for each direction L and V

Reply:
I've proposed optimized variant called `shadingNormal_masking` in https://github.com/Devsh-Graphics-Programming/Nabla/issues/310 but it should be quadruple checked